File: src/api/lib/blue-brain-training/schema.go
// KONTROL BLUE AI — COGNITIVE LEARNING REGISTRY (GO MODEL)
// Path: /src/api/lib/blue-brain-training/schema.go
// Implements concurrent-safe learning schemas and signature verification for Go backend nodes.

package training

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"time"
)

type TrainingSource string

const (
	SystemAutomated TrainingSource = "SYSTEM"
	UserFeedback    TrainingSource = "USER_FEEDBACK"
	ExpertInjection TrainingSource = "EXPERT_INJECTION"
)

// BlueBrainTrainingPair represents any learned pattern vector
type BlueBrainTrainingPair struct {
	ID           string         `json:"id" db:"id"`
	Prompt       string         `json:"prompt" db:"prompt"`
	Response     string         `json:"response" db:"response"`
	Category     string         `json:"category" db:"category"`
	Source       TrainingSource `json:"source" db:"source"`
	Confidence   float64        `json:"confidence" db:"confidence"`
	SecurityHash string         `json:"security_hash" db:"security_hash"`
	CreatedAt    int64          `json:"created_at" db:"created_at"`
}

// CognitiveIndex gauges current real-time data sync indicators
type CognitiveIndex struct {
	ID            string `json:"id" db:"id"`
	ModuleKey     string `json:"module_key" db:"module_key"`
	IndexName     string `json:"index_name" db:"index_name"`
	RecordCount   int    `json:"record_count" db:"record_count"`
	LastIndexedAt int64  `json:"last_indexed_at" db:"last_indexed_at"`
}

// SecurityShield provides secure validation routines for training collections
type SecurityShield struct{}

// VerifySignature validates a learning frame's signature to block untrusted inputs
func (s *SecurityShield) VerifySignature(pair *BlueBrainTrainingPair, secretKey string) bool {
	h := hmac.New(sha256.New, []byte(secretKey))
	
	// Create secure validation message representation
	message := fmt.Sprintf("%s:%s:%s:%s", pair.ID, pair.Prompt, pair.Response, pair.Source)
	h.Write([]byte(message))
	
	expectedSignature := hex.EncodeToString(h.Sum(nil))
	
	// Safe byte comparison to deter timing attacks
	return hmac.Equal([]byte(pair.SecurityHash), []byte(expectedSignature))
}

File: src/api/lib/blue-brain-training/schema.rs
// KONTROL BLUE AI — COGNITIVE LEARNING REGISTRY (RUST MODEL)
// Path: /src/api/lib/blue-brain-training/schema.rs
// Implements secure, memory-safe, and immutable learning nodes for Rust-based system pipelines.

use serde::{Serialize, Deserialize};
use std::time::SystemTime;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum TrainingSource {
    SystemAutomated,
    UserFeedback,
    ExpertInjection,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BlueBrainTrainingPair {
    pub id: String,
    pub prompt: String,
    pub response: String,
    pub category: String,
    pub source: TrainingSource,
    pub confidence: f64,
    pub security_hash: String, // Validates payload with Rust cryptographic signature
    pub created_at: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CognitiveIndex {
    pub id: String,
    pub module_key: String,
    pub index_name: String,
    pub record_count: u32,
    pub last_indexed_at: u64,
}

pub struct SecurityShield;

impl SecurityShield {
    /// Validates the training pair integrity using a secure cryptographic SHA256 checksum
    /// preventing malicious injections into Blue's core long-term neural memory.
    pub fn verify_signature(pair: &BlueBrainTrainingPair, secret_key: &str) -> bool {
        use hmac::{Hmac, Mac};
        use sha2::Sha256;

        type HmacSha256 = Hmac<Sha256>;

        let mut mac = HmacSha256::new_from_slice(secret_key.as_bytes())
            .expect("HMAC keys can be of any size");
        
        let payload = format!("{}:{}:{}:{:?}", pair.id, pair.prompt, pair.response, pair.source);
        mac.update(payload.as_bytes());
        
        let result = mac.finalize();
        let expected_bytes = result.into_bytes();
        let expected_hex = hex::encode(expected_bytes);

        pair.security_hash == expected_hex
    }
}
